src/monotonic_path_preview/mod.rs

```diff
@@ -0,0 +1,911 @@
+    /// is a symlink to `/example/path/b`, and `b` is a symlink back to `a`,
+    /// then giving this method a path like `/example/path/a/../c` will return
+    /// this error. It's like the POSIX `ELOOP` error, but cross-platform.
+    ///
+    /// Note that "does not exist" is *not* a fatal error for this function;
+    /// path components that do not exist by definition are not symlinks, and
+    /// are treated the same way as every other component that is not a symlink.
+    ///
+    /// [`Error::IoError`]: struct.Error.html#variant.IoError
+    /// [`Error::SymlinkLoops`]: struct.Error.html#variant.SymlinkLoops
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # fn example() -> Result<(), Error> {
+    /// let real_current_directory = Absolute::new(std::env::current_dir()?)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new<P: AsRef<path::Path>>(path: P) -> Result<Absolute, Error> {
+        // This function steps through the components of path, checking each one
+        // for monotonicity and building up a proper monotonic path in `res`.
+        // `tail` contains the path components we've yet to check and clean up.
+        let (mut res, mut tail) = split_head_and_tail(path)?;
+
+        let mut seen_paths = collections::BTreeSet::new();
+
+        // While there's still components left to check...
+        while tail.len() > 0 {
+            debug!(
+                "Monotonic path: {:?}, to check: {:?}",
+                res, tail
+            );
+
+            // Grab the component at the front of the tail, so we can check it.
+            let current = tail.pop_front()
+                .expect("len() > 0 but vec is empty?");
+
+            if &current == "." {
+                // A "." component can be ignored.
+                //
+                //    blah/./blah -> blah/blah
+
+            } else if &current == ".." {
+                // If we get a ParentDir component, the component at the end of
+                // `res` might be a symlink. in which case we'll have to splice
+                // the target path in at the beginning of `tail` so we have the
+                // full, monotonic path.
+                //
+                // If `foo` is not a symlink:
+                //
+                //    blah/foo/../blah -> blah/blah
+                //
+                // If `foo` is a symlink to `relative/target`:
+                //
+                //    blah/symlink/../blah -> blah/relative/target/../blah
+                //
+                // If `foo` is a symlink to `/absolute/target`:
+                //
+                //    blah/symlink/../blah -> /absolute/target/../blah
+                //
+
+                // But first, let's check for symlink loops. If we've already
+                // dereferenced this symlink before, we've hit a loop and we
+                // will never find a suitable answer.
+                if seen_paths.contains(&res) {
```

Comment:
this can be done in one operation:

```
if let Some(dup) = seen_paths.insert(res.clone()) {
    return Err(... dup);
}
```

Reply:
That does mean we always call `res.clone()` even if we don't need to, but I guess it's not a waste in the common case (adding a never-before-seen path to `seen_paths`), so the brevity is worth it.